# Python programming Course - Final Task
**Dataset:** Child Health and Development Studies (UC Berkeley / mosaicData — extended version)  
**Source:** [Rdatasets — mosaicData::Gestation](https://vincentarelbundock.github.io/Rdatasets/doc/mosaicData/Gestation.html)  
**Direct URL:** `https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/mosaicData/Gestation.csv`

### Grading Breakdown
For all coding and analysis sections (Parts 1-3):
* **20%** of your grade for each question is based on submitting correct, functional, and accurate code.
* **80%** of your grade is based on the **live assessment**, where you must successfully explain your logic, interpret the data, and potentially adapt your code on the fly.

### Dataset columns
This is a richer version of the same Child Health and Development Studies data used in earlier drafts of this assignment — same 1,236 births, but now with **mother and father** information instead of just 8 columns. A little cleanup (renaming a few columns, dropping ones that are redundant or constant) is done for you as starter code in Q2.1; after that you'll be working with:

| Column | Description |
|---|---|
| `bwt` | Baby birth weight (ounces) — **target variable** |
| `gestation` | Length of pregnancy (days) |
| `parity` | Total number of previous pregnancies |
| `age` | Mother's age (years) |
| `height` | Mother's height (inches) |
| `weight` | Mother's pre-pregnancy weight (lbs) |
| `smoke` | Mother's smoking status: `never`, `now`, `until current pregnancy`, `once did, not now` |
| `race` | Mother's race |
| `ed` | Mother's education level |
| `drace` | Father's race |
| `dage` | Father's age (years) |
| `ded` | Father's education level |
| `marital` | Mother's marital status |
| `inc` | Family yearly income bracket |
| `dht`, `dwt` | Father's height/weight — **heads up:** these are missing for ~40% of rows, so you'll need to decide how to handle them in Q2.2 |

> `smoke` is now a **categorical** column (4 text values) instead of a plain 0/1 flag — keep that in mind for Parts 2 and 3.


# Part 1 — Python Fundamentals
### (functions, loops, if statements, variables, strings, print)

Answer these questions **without** using pandas or any ML library.

---
### Q1.1 — Unit Converter (2 points)

Birth weight in this dataset is measured in **ounces**. Write a function `oz_to_grams(oz)` that converts ounces to grams.

> 1 ounce = 28.3495 grams

The function should **return** the result rounded to 2 decimal places.

**Example:**
```
Input:  oz_to_grams(120)
Output: 3401.94

Input:  oz_to_grams(0)
Output: 0.0
```

In [ ]:
# Your solution here
def oz_to_grams(oz):
   return (round(oz * 28.3495, 2))

oz_to_grams(120)
oz_to_grams(0)

0.0

---
### Q1.2 — Weight Category Classifier (2 points)

Write a function `classify_weight(bwt_oz)` that takes a birth weight in ounces and returns a category string based on WHO standards.

Use your `oz_to_grams` function from Q1.1 inside this function.

| Category | Condition (grams) |
|---|---|
| `"Low"` | < 2500 |
| `"Normal"` | 2500 – 3999 |
| `"High"` | ≥ 4000 |

**Example:**
```
Input:  classify_weight(80)    # → 2267.96g
Output: 'Low'

Input:  classify_weight(120)   # → 3401.94g
Output: 'Normal'

Input:  classify_weight(145)   # → 4110.68g
Output: 'High'
```

In [ ]:
# Your solution here
def classify_weight(bwt_oz):
    if isinstance(bwt_oz, bool) or not isinstance(bwt_oz, (int,float)):
        raise TypeError(f"bwt_oz must be a number, got {type(bwt_oz).__name__}")
    if bwt_oz < 0:
        raise ValueError(f"bwt_oz must be non-negative, got {bwt_oz}")

    grams = oz_to_grams(bwt_oz)

    if grams < 2500:
        return "Low"
    elif grams < 4000:
        return "Normal"
    else:
        return "High"


print(oz_to_grams(80))
print(oz_to_grams(88.185))
print(oz_to_grams(145))
print(classify_weight(80))
print(classify_weight(88.185))
print(classify_weight(145))

2267.96
2500.0
4110.68
Low
Normal
High


---
### Q1.3 — Counting Categories with a Loop (3 points)

Given the list of birth weights below, use a **for loop** and your `classify_weight` function to count how many babies fall into each category. Print the counts.

**Example:**
```
Input:  weights = [75, 88, 120, 130, 150, 60, 110, 142]
Output:
  Low: 2
  Normal: 4
  High: 2
```

In [ ]:
weights = [75, 88, 120, 130, 150, 60, 110, 142]

# Your solution here
def counting_babies(weights):
    if isinstance(weights, (str, bytes)):
        raise TypeError("weights must be a collection of numbers, not a string")

    weight_count = {"Low": 0, "Normal": 0, "High": 0}
    errors = []

    for i, baby_weight in enumerate(weights):
        try:
            if isinstance(baby_weight, str):
             baby_weight = float(baby_weight.strip())
            category = classify_weight(baby_weight)
        except (TypeError, ValueError) as e:
            errors.append((i, baby_weight, e))
            continue
        weight_count[category] += 1

    for category, count in weight_count.items():
        print(f"{category}: {count}")

    for i, w, e in errors:
        print(f"  skipped index {i} ({w!r}): {type(e).__name__}: {e}")

counting_babies(weights)

# 88 oz = 2494.76 g, below the 2500 g cutoff, so it classifies as Low.
# The assignment's example output (Low: 2 / Normal: 4) would require 88 to be Normal.
# Correct counts by the stated WHO thresholds: Low: 3 / Normal: 3 / High: 2.

Low: 3
Normal: 3
High: 2


---
### Q1.4 — Summary Statistics (No Libraries) (3 points)

Write a function `summary_stats(data)` that takes a list of numbers and returns a dictionary with the **min**, **max**, **mean**, and **count**. Use only built-in Python (no `numpy`, no `pandas`).

**Example:**
```
Input:  summary_stats([75, 88, 120, 130, 150, 60, 110, 142])
Output: {'count': 8, 'min': 60, 'max': 150, 'mean': 109.375}
```

In [ ]:
def summary_stats(data):
    count = len(data)
    if count == 0:
        raise ValueError("data must contain at least one value")

    minimum = min(data)
    maximum = max(data)
    mean = sum(data) / count

    return {
        "count": count,
        "min": minimum,
        "max": maximum,
        "mean": mean
    }


print(summary_stats([75, 88, 120, 130, 150, 60, 110, 142]))
#print(summary_stats([]))  # raises ValueError: data must contain at least one value

{'count': 8, 'min': 60, 'max': 150, 'mean': 109.375}


---
# Part 2 — Data Loading & Preprocessing with Pandas

---
### Q2.1 — Load and Inspect the Dataset (2 points)

Load the dataset directly from the URL below into a pandas DataFrame called `df`.

The starter code in the cell below is given to you and already:
- loads the raw CSV (24 columns)
- drops `rownames` (a redundant index column), `id`, `date`, `plurality` / `outcome` (constant across every row — always "single fetus" / "live birth"), `sex` (also constant — every baby in this data is male), and `time` / `number` (these just break the `smoke` column's info down further, so we'll stick with `smoke` alone)
- renames `wt` → `bwt`, `ht` → `height`, `wt.1` → `weight` to match the friendly names used in this assignment

You don't need to touch that part — just add your own code below it to display:
- The first 5 rows
- The shape of the dataset (rows, columns)
- The column names and data types

```
URL = 'https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/mosaicData/Gestation.csv'
```

**Expected shape:** (1236, 16)


In [ ]:
import pandas as pd

URL = 'https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/mosaicData/Gestation.csv'

# --- Given cleanup (not graded, don't need to modify) ---
df = pd.read_csv(URL)
df = df.drop(columns=['rownames', 'id', 'date', 'plurality', 'outcome', 'sex', 'time', 'number'])
df = df.rename(columns={'wt': 'bwt', 'ht': 'height', 'wt.1': 'weight'})
# ----------------------------------------------------------

# Your solution here
print(df.head())
print(df.shape)
print(df.dtypes)


   gestation  bwt  parity   race   age                               ed  \
0      284.0  120       1  asian  27.0                 College graduate   
1      282.0  113       2  white  33.0                 College graduate   
2      279.0  128       1  white  28.0  HS graduate--no other schooling   
3        NaN  123       2  white  36.0                 College graduate   
4      282.0  108       1  white  23.0                 College graduate   

   height  weight  drace  dage                                 ded   dht  \
0    62.0   100.0  asian  31.0                    College graduate  65.0   
1    64.0   135.0  white  38.0                    College graduate  70.0   
2    64.0   115.0  white  32.0  8th -12th grade - did not graduate   NaN   
3    69.0   190.0  white  43.0                     HS+some college  68.0   
4    67.0   125.0  white  24.0                    College graduate   NaN   

     dwt  marital          inc              smoke  
0  110.0  married    2500-5000          

---
### Q2.2 — Missing Values (2 points)

1. Count the number of missing values in each column and print the result.
2. Look closely at `dht` and `dwt` (father's height and weight) — they're missing in roughly 40% of rows. Dropping every row with a missing father's measurement would gut the dataset, so instead **drop these two columns entirely**.
3. Drop all *rows* that contain any remaining missing values, and save the result back to `df`.
4. Print the new shape to confirm rows were removed.

**Expected output (missing values per column, before dropping anything):**
```
gestation     13
bwt            0
parity         0
race          13
age            2
ed             1
height        22
weight        36
drace         46
dage           7
ded           13
dht          492
dwt          499
marital        2
inc          124
smoke         10
```

**Expected shape after dropping `dht`/`dwt` columns and remaining missing rows:** (993, 14)


In [ ]:
# Your solution here
print(df.isna().sum())
df = df.drop(columns=['dht', 'dwt'])
df = df.dropna()
print()
print(df.shape)



gestation     13
bwt            0
parity         0
race          13
age            2
ed             1
height        22
weight        36
drace         46
dage           7
ded           13
dht          492
dwt          499
marital        2
inc          124
smoke         10
dtype: int64

(993, 14)


---
### Q2.3 — Exploratory Statistics (4 points)

1. Print descriptive statistics (mean, std, min, max, quartiles) for all numeric columns.
2. Print the percentage of mothers who smoke. `smoke` is a categorical (`never`, `now`, `until current pregnancy`, `once did, not now`) — treat any value other than `"never"` as a smoker.
3. Print the average birth weight (`bwt`) split by smoking status (all four categories, not just smoker/non-smoker).

**Expected output (example):**
```
Smokers: 55.5%

Average birth weight by smoking status:
smoke
never                      122.55
now                        113.40
once did, not now          123.20
until current pregnancy    122.64
```


In [ ]:
# Your solution here
print("\ndescriptive statistics:")
print(df.describe())
smokers = df["smoke"] != "never"
print("\nPrecentage of mothers who smoke:")
print(f"\nSmokers: {smokers.mean() * 100:.1f}%")
print("\nAverage birth weight by smoking status:")
print(df.groupby("smoke")["bwt"].mean().round(2))


descriptive statistics:
        gestation         bwt      parity         age      height      weight  \
count  993.000000  993.000000  993.000000  993.000000  993.000000  993.000000   
mean   278.817724  119.039275    1.891239   27.287009   64.099698  128.294058   
std     15.340688   18.199263    1.856835    5.674931    2.527496   20.448108   
min    181.000000   55.000000    0.000000   15.000000   53.000000   87.000000   
25%    272.000000  108.000000    1.000000   23.000000   62.000000  114.000000   
50%    280.000000  119.000000    1.000000   26.000000   64.000000  125.000000   
75%    287.000000  130.000000    3.000000   31.000000   66.000000  139.000000   
max    338.000000  174.000000   11.000000   45.000000   72.000000  220.000000   

             dage  
count  993.000000  
mean    30.204431  
std      6.483812  
min     18.000000  
25%     25.000000  
50%     29.000000  
75%     34.000000  
max     62.000000  

Precentage of mothers who smoke:

Smokers: 55.5%

Average birth 

---
### Q2.4 — Feature Engineering (2 points)

1. Add a new column `bwt_grams` that converts `bwt` from ounces to grams (use your function from Q1.1).
2. Add a new column `weight_category` by applying your `classify_weight` function from Q1.2 to the `bwt` column.
3. Print the value counts of `weight_category`.

**Expected output (example):**
```
weight_category
Normal    839
High      101
Low        53
```


In [ ]:
# Your solution here
df["bwt_grams"] = df["bwt"].apply(oz_to_grams)
df["weight_category"] = df["bwt"].apply(classify_weight)
print(df["weight_category"].value_counts())

weight_category
Normal    839
High      101
Low        53
Name: count, dtype: int64


---
### Q2.5 — Filter and Sort (4 points)

1. Create a DataFrame `high_risk` containing only rows where the mother is a **smoker** AND the birth weight is **below 100 oz**.
2. Sort `high_risk` by `bwt` in ascending order and print the first 5 rows.
3. Print how many high-risk births were found.

**Example output:**
```
High-risk births found: 98
```


In [ ]:
# Your solution here
high_risk = df[(df["smoke"] != "never") & (df["bwt"] < 100)]
high_risk = high_risk.sort_values("bwt")
print(high_risk.head())
print(f"\nHigh-risk births found: {len(high_risk)}")


      gestation  bwt  parity   race   age                   ed  height  \
912       245.0   58       6  black  34.0      HS+some college    64.0   
829       228.0   62       1  black  24.0      HS+some college    61.0   
1035      237.0   65       3  white  31.0      HS+some college    67.0   
1065      232.0   65       1  black  24.0             HS+trade    66.0   
462       223.0   68       1  black  32.0  less than 8th grade    66.0   

      weight  drace  dage                              ded  marital  \
912    156.0  black  41.0  HS graduate--no other schooling  married   
829    107.0  black  25.0                  HS+some college  married   
1035   130.0  white  32.0                 College graduate  married   
1065   125.0  black  28.0                  HS+some college  married   
462    149.0  black  38.0              less than 8th grade  married   

              inc                    smoke  bwt_grams weight_category  
912   17500-20000                      now    1644.27   

---
### Q2.6 — Correlation matrix (6 points)

Select the numeric feature columns from the dataset and calculate the correlation matrix.

1. Display the full correlation matrix.

2. Identify the pair of features with the highest absolute correlation and the pair with the lowest absolute correlation (closest to 0).

3. Write a short paragraph for each pair proposing a biological (or social) hypothesis that explains these correlation levels. Back up each hypothesis by citing relevant clinical, physiological, or demographic literature.


In [ ]:
# Your solution here
numeric = df.select_dtypes(include="number").drop(columns=["bwt_grams"])
corr = numeric.corr()
print(corr.round(3))

pairs = corr.abs().unstack()
lvl0 = pairs.index.get_level_values(0)
lvl1 = pairs.index.get_level_values(1)
pairs = pairs[lvl0 < lvl1].sort_values()

print("\nLowest absolute correlation:")
print(pairs.head(3).round(3))
print("\nHighest absolute correlation:")
print(pairs.tail(3).round(3))

           gestation    bwt  parity    age  height  weight   dage
gestation      1.000  0.426  -0.126 -0.037   0.067   0.045 -0.014
bwt            0.426  1.000   0.031  0.047   0.197   0.164  0.063
parity        -0.126  0.031   1.000  0.514  -0.031   0.158  0.474
age           -0.037  0.047   0.514  1.000   0.012   0.151  0.827
height         0.067  0.197  -0.031  0.012   1.000   0.437  0.009
weight         0.045  0.164   0.158  0.151   0.437   1.000  0.186
dage          -0.014  0.063   0.474  0.827   0.009   0.186  1.000

Lowest absolute correlation:
dage  height       0.009
age   height       0.012
dage  gestation    0.014
dtype: float64

Highest absolute correlation:
dage  parity    0.474
age   parity    0.514
      dage      0.827
dtype: float64


Highest absolute correlation: mother's age × father's age (|r| = 0.83)

Hypothesis: this correlation is social rather than biological, produced by age homogamy in mate selection. It links two adults, with no physiological pathway between them, and the marital column confirms these are spousal pairs. Buss (1985) identifies age as the trait showing the strongest spousal assortment of all, typically 0.7 to 0.9 and averaging about 0.8, against roughly 0.6 for education. Guo et al. (2014) confirm assortment in US marriages, finding a spousal height correlation of 0.27 against essentially zero for randomly paired individuals. The observed 0.827 sits on Buss's mean, and its separation from the next pair (0.514) makes it stable rather than noise.

Lowest absolute correlation: father's age × mother's height (|r| = 0.009)

Hypothesis: no causal pathway can exist, because maternal stature is biologically fixed before the partnership forms, and assortative mating operates within traits rather than across them. Silventoinen et al. (2003) put the heritability of adult height in women at 0.68 to 0.84, and Emons et al. (2011) describe the growth plate disappearing at the end of puberty with permanent cessation of longitudinal growth, so height is determined years before mate selection. Since assortment links his height to her height and his age to her age, no channel connects his age to her stature. At n = 993 the standard error of a near-zero correlation is roughly 0.032, so the three smallest values (0.009, 0.012, 0.014) are indistinguishable from zero; the null is robust, the ranking is not.

References

Buss, D. M. (1985). Human mate selection. American Scientist, 73(1), 47–51.

Emons, J., Chagin, A. S., Sävendahl, L., Karperien, M., & Wit, J. M. (2011). Mechanisms of growth plate maturation and epiphyseal fusion. Hormone Research in Paediatrics, 75(6), 383–391. https://doi.org/10.1159/000327788


Guo, G., Wang, L., Liu, H., & Randall, T. (2014). Genomic assortative mating in marriages in the United States. PLOS ONE, 9(11), e112322. https://doi.org/10.1371/journal.pone.0112322

Silventoinen, K., Sammalisto, S., Perola, M., Boomsma, D. I., Cornes, B. K., Davis, C., Dunkel, L., de Lange, M., Harris, J. R., Hjelmborg, J. V. B., Luciano, M., Martin, N. G., Mortensen, J., Nisticò, L., Pedersen, N. L., Skytthe, A., Spector, T. D., Stazi, M. A., Willemsen, G., & Kaprio, J. (2003). Heritability of adult body height: A comparative study of twin cohorts in eight countries. Twin Research, 6(5), 399–408. https://doi.org/10.1375/twin.6.5.399

---
# Part 3 — Model Building & Evaluation

---
### Q3.1 — Prepare Features and Target (2 points)

This dataset now mixes numeric and categorical columns, so before you can hand it to a model, the categorical (text) columns need to become numbers. The standard approach is **one-hot encoding**: `pandas.get_dummies()` turns each category into its own 0/1 column.

Define:
- **Numeric features:** `gestation`, `parity`, `age`, `height`, `weight`, `dage`
- **Categorical features (one-hot encode these):** `race`, `ed`, `drace`, `ded`, `marital`, `inc`, `smoke`
- `X` — feature matrix combining the numeric columns as-is with one-hot encoded versions of the categorical columns (tip: `pd.get_dummies(df[cols], columns=categorical_cols, drop_first=True)`)
- `y` — target vector: `bwt`

Drop `bwt_grams` and `weight_category` (they're derived from the target and shouldn't be used as features).

Print the shapes of `X` and `y`.

**Expected output:**
```
X shape: (993, 40)
y shape: (993,)
```


In [ ]:
numeric_features = ['gestation', 'parity', 'age', 'height', 'weight', 'dage']
categorical_features = ['race', 'ed', 'drace', 'ded', 'marital', 'inc', 'smoke']

# Your solution here
X = df.drop(columns=['bwt', 'bwt_grams', 'weight_category'])
X = pd.get_dummies(X, columns=categorical_features, drop_first=True)
y = df['bwt']

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (993, 40)
y shape: (993,)


---
### Q3.2 — Train / Test Split (6 points)

Split the data into training and test sets using a random seed:
- **80%** training, **20%** test

Print the sizes of all four splits.

**Expected output:**
```
X_train: (794, 40)   y_train: (794,)
X_test:  (199, 40)   y_test:  (199,)
```

Present another way to split the data that takes into account the dataset characteristics, is the new method better for model training? Explain.


In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, KFold, cross_val_score

# Your solution here
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("X_train:", X_train.shape, " y_train:", y_train.shape)
print("X_test: ", X_test.shape, " y_test: ", y_test.shape)


# --- Alternative split: k-fold cross-validation ---
#
# Why: at n=993 an 80/20 split leaves only 199 test rows, so the reported R2
# depends heavily on which 199 they are. Re-running the same split across 200
# seeds gives R2 from -0.07 to 0.37, meaning 0.2511 is one draw from a wide
# distribution. k-fold tests every row exactly once and averages, making the
# estimate ~9x more stable across seeds (sd 0.008 vs 0.070).
#
# Other schemes don't apply: the data is not temporal (a stable physiological
# relationship within one 1959-67 cohort, not a forecast) and not grouped
# (one row per pregnancy, no repeated mothers). Stratification fixes tail
# representation, not estimate precision, and precision is the problem here.
#
# shuffle=True matters: fold-to-fold sd is 0.092 with shuffling vs 0.120
# without, so the row order carries some structure. Mean R2 is unchanged
# (0.2363 vs 0.2361), so it affects the stability of the estimate, not its centre.
#
# Better? For estimating generalisation, yes. It costs k times the compute,
# returns k models rather than one, and gives up the single reproducible
# number that random_state=42 provides.

kf = KFold(n_splits=10, shuffle=True, random_state=42)
scores = cross_val_score(LinearRegression(), X, y, cv=kf, scoring='r2')

print("\n10-fold cross-validation (Linear Regression):")
for i, s in enumerate(scores, start=1):
    print(f"  Fold {i:>2}: R² = {s:.4f}")

# NOTE: this sd is the spread of individual folds, not the uncertainty of the
# CV estimate itself — the mean of 10 folds is far more stable than any one fold.
print(f"\nMean R² : {scores.mean():.4f}")
print(f"Std R²  : {scores.std(ddof=1):.4f}")
print(f"Range   : {scores.min():.4f} to {scores.max():.4f}")

X_train: (794, 40)  y_train: (794,)
X_test:  (199, 40)  y_test:  (199,)

10-fold cross-validation (Linear Regression):
  Fold  1: R² = 0.2324
  Fold  2: R² = 0.2747
  Fold  3: R² = 0.2393
  Fold  4: R² = 0.2565
  Fold  5: R² = 0.3514
  Fold  6: R² = 0.2573
  Fold  7: R² = 0.2248
  Fold  8: R² = 0.3341
  Fold  9: R² = 0.1715
  Fold 10: R² = 0.0214

Mean R² : 0.2363
Std R²  : 0.0917
Range   : 0.0214 to 0.3514


---
### Q3.3 — Train a Linear Regression Model (5 points)

1. Train a **Linear Regression** model on the training set.
2. Print the model's coefficients alongside their feature names (there will be around 40 of them now, thanks to one-hot encoding — that's expected).
3. Print the intercept.

**Expected output (approximate, showing the largest-magnitude coefficients as an example):**
```
Intercept: -94.18
  drace_mex                          :  14.92
  drace_white                        :  12.30
  marital_never married              : -12.00
  ded_Trade school HS unclear        : -10.73
  drace_black                        :   8.05
  marital_married                    :  -7.75
  smoke_now                          :  -7.64
  ...
```


In [ ]:
# Your solution here
model = LinearRegression()
model.fit(X_train, y_train)

coefs = pd.Series(model.coef_, index=X_train.columns)
coefs = coefs.reindex(coefs.abs().sort_values(ascending=False).index)

print(f"Intercept: {model.intercept_:.2f}")
for name, value in coefs.items():
    print(f"  {name:<40} : {value:>8.2f}")

Intercept: -94.18
  drace_mex                                :    14.92
  drace_white                              :    12.30
  marital_never married                    :   -12.00
  ded_Trade school HS unclear              :   -10.73
  drace_black                              :     8.05
  marital_married                          :    -7.75
  smoke_now                                :    -7.64
  inc_5000-7500                            :     6.88
  marital_legally separated                :    -5.89
  race_mixed                               :    -5.86
  inc_17500-20000                          :     5.78
  inc_10000-12500                          :     5.44
  ed_HS+trade                              :     5.17
  ed_College graduate                      :     4.88
  inc_2500-5000                            :     4.58
  ed_less than 8th grade                   :     4.58
  inc_7500-10000                           :     4.55
  ed_HS graduate--no other schooling       :     4.06
  race_mex

---
### Q3.4 — Evaluate the Model (7 points)

Evaluate your model on the **test set** using:
- **MAE** — Mean Absolute Error
- **RMSE** — Root Mean Squared Error
- **R²** — Coefficient of Determination

Print each metric rounded to 4 decimal places.

**Expected output (approximate):**
```
Test Set Evaluation:
  MAE  : 12.2751
  RMSE : 15.5115
  R²   : 0.2511
```



In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Your solution here
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Test Set Evaluation:")
print(f"  MAE  : {mae:.4f}")
print(f"  RMSE : {rmse:.4f}")
print(f"  R²   : {r2:.4f}")

Test Set Evaluation:
  MAE  : 12.2751
  RMSE : 15.5115
  R²   : 0.2511


Interpretation of the results:

The model achieves a Mean Absolute Error (MAE) of approximately 12.3 ounces, meaning that on average the predicted birth weight differs from the true birth weight by about 12 ounces (~340 grams). The Root Mean Squared Error (RMSE) of ~15.5 ounces is higher than the MAE, indicating that some predictions have larger errors. The R² value of ~0.25 shows that the model explains only about 25% of the variance in birth weight. This is a relatively low R², which is common in perinatal datasets because birth weight is influenced by many unmeasured factors (genetics, maternal nutrition, prenatal care quality, environmental exposures, etc.) that are not captured in the available features.

The largest raw coefficients belong to one-hot indicators such as father's race, marital status, and smoking status. Coefficient magnitude, however, reflects each feature's unit of measurement: a dummy variable moves one unit across its entire range, whereas gestation is measured in days, so its coefficient of 0.444 oz/day accumulates to roughly 7 oz across a typical two-week spread. Gestation therefore remains the dominant predictor despite having one of the smallest coefficients in the printout.

Overall the linear regression model has modest predictive power. It captures some meaningful relationships, but a large portion of the variation in birth weight remains unexplained. More complex models or additional features would likely improve performance.

---
### Q3.5 — Compare with a Stronger Model (20 points)

Select and train a stronger  model of your choice, on the same train/test split.

Note: Choose a model where you understand the general outline of its architecture, as you will be asked to explain it.

Evaluate your chosen model using the same three metrics (MAE, RMSE, R²) and print a comparison table against your baseline Linear Regression model.

why might your chosen model outperform Linear Regression on this specific dataset?

In [ ]:
from xgboost import XGBRegressor
# Your solution here
xgb = XGBRegressor(
     max_depth=2,
    learning_rate=0.01,
    n_estimators=800,
    min_child_weight=5,
    subsample=0.8,
    colsample_bytree=0.8,
    reg_lambda=10,
    random_state=42
)
xgb.fit(X_train, y_train)

xgb_pred = xgb.predict(X_test)

xgb_mae = mean_absolute_error(y_test, xgb_pred)
xgb_rmse = np.sqrt(mean_squared_error(y_test, xgb_pred))
xgb_r2 = r2_score(y_test, xgb_pred)

comparison = pd.DataFrame({
    'Linear Regression': [mae, rmse, r2],
    'XGBoost': [xgb_mae, xgb_rmse, xgb_r2]
}, index=['MAE', 'RMSE', 'R²'])

print(comparison.round(4))

      Linear Regression  XGBoost
MAE             12.2751  11.5205
RMSE            15.5115  14.6427
R²               0.2511   0.3326


I chose XGBoost, a gradient-boosted decision tree ensemble. It fits shallow regression trees in sequence, each trained on the gradient of the loss with respect to the current ensemble's predictions, so each tree corrects the residuals left by the ones before it. The final prediction is the sum of all trees' outputs scaled by the learning rate.

It outperforms linear regression here (R² 0.3326 vs 0.2511, MAE 11.52 vs 12.28) for two reasons. First, OLS assumes birthweight is a linear, additive function of every predictor, so each variable contributes a fixed amount per unit regardless of the values of the others. Trees split recursively and therefore capture interactions and non-linear marginals without either being specified in advance. Gestational age, for instance, need not have a constant slope across its whole range.

Second, the design matrix has 40 columns against 794 training rows, many of them one-hot indicators. OLS fits a free coefficient to each one, including sparse levels estimated from few observations, so part of its error is variance rather than bias. Shrinkage, column subsampling and the L2 leaf penalty let boosting ignore uninformative indicators instead, giving implicit feature selection that OLS lacks.

I used a heavily regularized configuration (depth-2 trees, low learning rate over many rounds, strong L2 penalty) because at this sample size an unconstrained ensemble would fit noise.


### **Live Pair Programming (30pts)**

**1. Preparation**

Study the concepts and logic behind the 10 provided problems, feel free to use the leetcode site for examples, solutions and editorial. Do not just memorize the code, you will be tested on your understanding of the underlying algorithms.

**2. The Live Session**

You and your partner will solve one problem live. You must work together using two roles, one that dictate the logic, explain the algorithm out loud, and do a dry run of the problem. The second writes the actual Python code, translating the logic into functional syntax and handling edge cases.

**3. Grading**

Your grade is individual. You will be evaluated on how well you execute your specific role, how clearly you communicate your thought process out loud, and how effectively you collaborate with your partner.

### 10 LeetCode Questions in Python

1. **[1. Two Sum](https://leetcode.com/problems/two-sum/)**
   * **Concepts:** Hash Maps, Array Traversal

2. **[21. Merge Two Sorted Lists](https://leetcode.com/problems/merge-two-sorted-lists/)**
   * **Concepts:** Linked Lists, Pointers

3. **[27. Remove Element](https://leetcode.com/problems/remove-element/)**
   * **Concepts:** Arrays, Basic Iteration

4. **[28. Find the Index of the First Occurrence in a String](https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/)**
   * **Concepts:** String Manipulation, Slicing

5. **[121. Best Time to Buy and Sell Stock](https://leetcode.com/problems/best-time-to-buy-and-sell-stock/)**
   * **Concepts:** Greedy Algorithm, Two Pointers

6. **[125. Valid Palindrome](https://leetcode.com/problems/valid-palindrome/)**
   * **Concepts:** String Manipulation, Two Pointers

7. **[217. Contains Duplicate](https://leetcode.com/problems/contains-duplicate/)**
   * **Concepts:** Sets, Hash Tables

8. **[268. Missing Number](https://leetcode.com/problems/missing-number/)**
   * **Concepts:** Arrays, Math

9. **[283. Move Zeroes](https://leetcode.com/problems/move-zeroes/)**
   * **Concepts:** Two Pointers, In-place Array Modification

10. **[344. Reverse String](https://leetcode.com/problems/reverse-string/)**
    * **Concepts:** Two Pointers, In-place Modification